### PR TITLE
Update idle, newapp, newbuild cmds

### PR DESCRIPTION
**Pull request by juanvallejo**
Updates the idle, newapp, and newbuild commands to use upstream patterns.
Picks upstream https://github.com/kubernetes/kubernetes/pull/66131

cc @deads2k

### CODE REVIEW

File: pkg/oc/cli/cmd/newapp.go [hunk 1]
```diff
@@ -223,59 +257,56 @@ func NewCmdNewApplication(name, baseName string, f kcmdutil.Factory, streams gen
...
+	cmd.Flags().StringSliceVarP(&o.Config.ImageStreams, "image", "", o.Config.ImageStreams, "Name of an image stream to use in the app. (deprecated)")
 	cmd.Flags().MarkDeprecated("image", "use --image-stream instead")
-	cmd.Flags().StringSliceVarP(&config.ImageStreams, "image-stream", "i", config.ImageStreams, "Name of an image stream to use in the app.")
-	cmd.Flags().StringSliceVar(&config.DockerImages, "docker-image", config.DockerImages, "Name of a Docker image to include in the app.")
-	cmd.Flags().StringSliceVar(&config.Templates, "template", config.Templates, "Name of a stored template to use in the app.")
-	cmd.Flags().StringSliceVarP(&config.TemplateFiles, "file", "f", config.TemplateFiles, "Path to a template file to use for the app.")
+	cmd.Flags().StringSliceVarP(&o.Config.ImageStreams, "image-stream", "i", o.Config.ImageStreams, "Name of an image stream to use in the app.")
+	cmd.Flags().StringSliceVar(&o.Config.DockerImages, "docker-image", o.Config.DockerImages, "Name of a Docker image to include in the app.")
+	cmd.Flags().StringSliceVar(&o.Config.Templates, "template", o.Config.Templates, "Name of a stored template to use in the app.")
```
Comment (juanvallejo) [inline pkg/oc/cli/cmd/newapp.go]: @deads2k I had to disable binding PrintFlags.TemplatePrinterFlags in this command (above) as it is overshadowed by this existing flag

**Comment by juanvallejo** [comment]
cc @deads2k

File: pkg/oc/cli/cmd/idle.go [hunk 2]
```diff
@@ -117,14 +123,32 @@ func (o *IdleOptions) Complete(f kcmdutil.Factory, cmd *cobra.Command, args []st
...
+	if err != nil {
+		return err
+	}
+
+	o.Mapper, err = f.ToRESTMapper()
+	if err != nil {
+		return err
+	}
+
+	o.Typer = legacyscheme.Scheme
```
Comment (deads2k) [inline pkg/oc/cli/cmd/idle.go]: This is likely to fail on non-groupified origin resources.  Use the read scheme we made.

Comment (deads2k) [inline pkg/oc/cli/cmd/idle.go]: why is this configurable?

File: pkg/oc/cli/cmd/idle.go [hunk 3]
```diff
@@ -244,12 +257,12 @@ func (o *IdleOptions) calculateIdlableAnnotationsByService(f kcmdutil.Factory) (
...
 		if !ok {
 			var mapping *meta.RESTMapping
-			mapping, err = mapper.RESTMapping(schema.GroupKind{Group: gv.Group, Kind: ref.Kind}, "")
+			mapping, err = o.Mapper.RESTMapping(schema.GroupKind{Group: gv.Group, Kind: ref.Kind}, "")
 			if err != nil {
 				return nil, err
 			}
 			var client resource.RESTClient
-			client, err = f.ClientForMapping(mapping)
+			client, err = o.ClientForMappingFn(mapping)
```
Comment (deads2k) [inline pkg/oc/cli/cmd/idle.go]: issue to remove this.  it will lead to sadness.

File: pkg/oc/cli/cmd/newapp.go [hunk 4]
```diff
@@ -138,34 +138,53 @@ To search templates, image streams, and Docker images that match the arguments p
 
 type ObjectGeneratorOptions struct {
 	Action bulk.BulkAction
+	genericclioptions.IOStreams
```
Comment (deads2k) [inline pkg/oc/cli/cmd/newapp.go]: nit, bottom of struct

File: pkg/oc/cli/cmd/newapp.go [hunk 5]
```diff
@@ -138,34 +138,53 @@ To search templates, image streams, and Docker images that match the arguments p
...
 	*ObjectGeneratorOptions
 }
 
+type versionedPrintObj struct {
+	printer printers.ResourcePrinter
+	cmd     *cobra.Command
+}
+
+func (p *versionedPrintObj) PrintObj(obj runtime.Object, out io.Writer) error {
```
Comment (deads2k) [inline pkg/oc/cli/cmd/newapp.go]: blech

File: pkg/oc/cli/cmd/newapp.go [hunk 6]
```diff
@@ -138,34 +138,53 @@ To search templates, image streams, and Docker images that match the arguments p
...
+	printFn := print.VersionedPrintObject(func(cmd *cobra.Command, obj runtime.Object, out io.Writer) error {
+		return p.printer.PrintObj(obj, out)
```
Comment (deads2k) [inline pkg/oc/cli/cmd/newapp.go]: I'll take it as a stepping stone to get the rest

File: pkg/oc/cli/cmd/newapp.go [hunk 7]
```diff
@@ -210,11 +230,29 @@ func (o *ObjectGeneratorOptions) Complete(baseName, commandName string, f kcmdut
...
-// NewCmdNewApplication implements the OpenShift cli new-app command.
-func NewCmdNewApplication(name, baseName string, f kcmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Command {
+func NewAppOptions(streams genericclioptions.IOStreams) *AppOptions {
 	config := newcmd.NewAppConfig()
 	config.Deploy = true
-	o := &NewAppOptions{&ObjectGeneratorOptions{Config: config}}
+
+	// disable the --template printFlag, as it is shadowed by the existing --template printing
+	// in this command - which is sued to select existing app templates
+	printFlags := genericclioptions.NewPrintFlags("")
```
Comment (deads2k) [inline pkg/oc/cli/cmd/newapp.go]: "created" I should think

File: pkg/oc/cli/cmd/newbuild.go [hunk 8]
```diff
@@ -84,91 +82,101 @@ on the Docker Hub.
...
-func NewCmdNewBuild(name, baseName string, f kcmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Command {
+func NewBuildOptions(streams genericclioptions.IOStreams) *BuildOptions {
 	config := newcmd.NewAppConfig()
 	config.ExpectToBuild = true
-	o := &NewBuildOptions{ObjectGeneratorOptions: &ObjectGeneratorOptions{Config: config}}
+
+	return &BuildOptions{
+		IOStreams: streams,
+		ObjectGeneratorOptions: &ObjectGeneratorOptions{
+			PrintFlags: genericclioptions.NewPrintFlags(""),
```
Comment (deads2k) [inline pkg/oc/cli/cmd/newbuild.go]: "created" I should think

File: pkg/oc/cli/cmd/newbuild.go [hunk 9]
```diff
@@ -84,91 +82,101 @@ on the Docker Hub.
...
+	cmd.Flags().BoolVar(&o.Config.AllowMissingImages, "allow-missing-images", o.Config.AllowMissingImages, "If true, indicates that referenced Docker images that cannot be found locally or in a registry should still be used.")
+	cmd.Flags().BoolVar(&o.Config.AllowMissingImageStreamTags, "allow-missing-imagestream-tags", o.Config.AllowMissingImageStreamTags, "If true, indicates that image stream tags that don't exist should still be used.")
+	cmd.Flags().StringVar(&o.Config.ContextDir, "context-dir", o.Config.ContextDir, "Context directory to be used for the build.")
+	cmd.Flags().BoolVar(&o.Config.NoOutput, "no-output", o.Config.NoOutput, "If true, the build output will not be pushed anywhere.")
+	cmd.Flags().StringVar(&o.Config.SourceImage, "source-image", o.Config.SourceImage, "Specify an image to use as source for the build.  You must also specify --source-image-path.")
+	cmd.Flags().StringVar(&o.Config.SourceImagePath, "source-image-path", o.Config.SourceImagePath, "Specify the file or directory to copy from the source image and its destination in the build directory. Format: [source]:[destination-dir].")
+
+	o.Action.BindForOutput(cmd.Flags(), "output", "template")
```
Comment (deads2k) [inline pkg/oc/cli/cmd/newbuild.go]: this looks weird.  Issue to scrub all this out.

**Comment by deads2k** [comment]
fairly minor comments.  fix, squash, self-tag

/approve

File: pkg/oc/cli/cmd/idle.go (hunk 3 above)
Comment (juanvallejo) [reply pkg/oc/cli/cmd/idle.go]: https://github.com/openshift/origin/issues/20312

File: pkg/oc/cli/cmd/newbuild.go (hunk 9 above)
Comment (juanvallejo) [reply pkg/oc/cli/cmd/newbuild.go]: https://github.com/openshift/origin/issues/20313

**Comment by juanvallejo** [comment]
@deads2k thanks, review comments addressed. Tagging

**Pull request merged**